storage-proofs-porep/Cargo.toml

```diff
@@ -19,7 +19,7 @@ merkletree = "0.21.0"
 mapr = "0.8.0"
 num-bigint = "0.2"
 num-traits = "0.2"
-sha2 = { version = "0.9.1", features = ["compress"] }
+sha2 = { version = "0.9.3", features = ["compress", "asm"] }
```

Comment:
We should not enable `"asm"` by default, this will result in build issues and compatibility accross other platforms

Reply:
Ah, good catch. Thanks!

Reply:
Thanks for the comments.

I run local tests with `lotus-bench` with this PR on AArch64 and AMD7742 to reproduce the issues as mentioned.
Lotus builds on both platforms are fine. On AMD7742, the x86 instrinsic backend, i.e. `sha2::sha256::x86::compress`, is used as before.

Could you please be more specific about the build and compatibility issues? I must miss something. :-）

Reply:
The issue is that the default flags are used to generate prebuilts, which sometimes end up not being portable anymore when `asm` is enabled by default. So instead this should be threaded through as a general `asm` feature and only be enabled in filecoin-ffi in the native/local builds.

Reply:
> filecoin-ffi

Sorry for being silent last few days, it's Spring festival.

Thanks for clarifying this.
I see we have some platform dependent code in this repository, like prefetch [1].
How do you think if we only enable `asm` feature for aarch64 here? Something like:

```
## rust-fil-proofs changes
diff --git a/storage-proofs/porep/Cargo.toml b/storage-proofs/porep/Cargo.toml
index e52f30d..f6549c4 100644
+[target."cfg(target_arch = \"aarch64\")".dependencies]
+sha2 = { version = "0.9.3", features = ["compress", "asm"] }
+[target."cfg(not(target_arch = \"aarch64\"))".dependencies]
+sha2 = { version = "0.9.3", features = ["compress"] }

##  filecoin-ffi changes
diff --git a/install-filcrypto b/install-filcrypto
index dac54b7..dc56a4d 100755
--- a/install-filcrypto
+++ b/install-filcrypto
@@ -150,12 +150,13 @@ build_from_source() {
     # Add feature specific rust flags as needed here.
+    additional_flags="-Zfeatures=itarget"
     if [ "${FFI_USE_BLST_PORTABLE}" == "1" ]; then
-        additional_flags="--no-default-features --features blst --features blst-portable"
+        additional_flags="--no-default-features --features blst --features blst-portable ${additional_flags}"
     elif [ "${FFI_USE_BLST}" == "1" ]; then
-        additional_flags="--no-default-features --features blst"
+        additional_flags="--no-default-features --features blst ${additional_flags}"
     else
-        additional_flags="--no-default-features --features pairing"
+        additional_flags="--no-default-features --features pairing ${additional_flags}"
     fi
```

[1] https://github.com/filecoin-project/rust-fil-proofs/blob/64390b6fcedb04dd1fdbe43c82b1e91c1439cea2/storage-proofs-porep/src/stacked/vanilla/macros.rs#L47

Reply:
yes I believe something like that should work

Reply:
OK, updated a commit. Only enable asm feature for aarch64 now.
Reopened https://github.com/filecoin-project/filecoin-ffi/pull/163 in `filecoin-ffi` for supporting conditional compilation of dependency feature based on target.
Thanks. :-)